internal/providers/docker/components/container/build.go

```diff
@@ -37,139 +35,134 @@ func (c *Container) canBuild() bool {
 }
 
 func (c *Container) buildImage(ctx context.Context) error {
-	buildContext, buildContextWriter := io.Pipe()
-	defer buildContextWriter.Close()
-	var eg errgroup.Group
-
 	buildTask := task.CurrentTask(ctx)
 	if buildTask == nil {
 		panic("No build task")
 	}
 
 	spec := c.Spec
 
-	eg.Go(func() error {
-		contextPath := filepath.Join(c.WorkspaceRoot, spec.Build.Context)
-		if !pathutil.HasFilePathPrefix(contextPath, c.WorkspaceRoot) {
-			return errors.New("docker container build context path must be in exo workspace root")
-		}
-		if err := tarBuildContext(buildContextWriter, contextPath); err != nil {
-			return fmt.Errorf("tarring build context: %w", err)
-		}
-		return nil
-	})
+	contextPath := filepath.Join(c.WorkspaceRoot, spec.Build.Context)
+	if !pathutil.HasFilePathPrefix(contextPath, c.WorkspaceRoot) {
+		return errors.New("docker container build context path must be in exo workspace root")
+	}
+	dockerfile := c.Spec.Build.Dockerfile
+	if dockerfile == "" {
+		dockerfile = "Dockerfile"
+	}
+	buildContext, err := getArchive(contextPath, dockerfile)
```

Comment:
This takes ownership of the reader, so you need to do a `defer buildContext.Close()`. If you don't things seem like they will be fine on the happy path, but if an error occurs reading from the archive, then the writing goroutine inside of `archive.TarWithOptions` will block indefinitely & this will be a goroutine leak.

Reply:
Good catch!